### DISCUSSION TITLE
Testing!

### DISCUSSION THREAD

**Discussion by TylerMatteo** [NYCPlanning/ae-zoning-api#118] [General]
Let's talk about how we can start to add tests to this repo! What we discuss here will inform our testing strategy on other Nest-based APIs in the future.

# Step 1: Read the docs
Nest has [a pretty good page in their fundamental section](https://docs.nestjs.com/fundamentals/testing) about how to do testing. It has examples of how to do unit testing as well as end-to-end testing. There are two dependencies used "under the hood" by Nest for testing out of the box:
* [Jest](https://github.com/jestjs/jest) for unit testing
* [Supertest](https://github.com/ladjs/supertest) for end to end

While both are equally important, I'm thinking we could start with unit tests? Nest's testing page shows an example of unit tests for a controller, but I think the overall gist would be similar to any Nest Provider (services and repositories are providers). @TangoYankee  and I have had some discussions about what really makes sense to test and what/how should be mocked at any given level of the architecture. Maybe as a first step, we create unit tests for one of our controllers, mocking any services that the controller needs. That would enable us to use Nest's example code as a 1 to 1 example. Thoughts?

**Comment by TangoYankee** [comment]
The value of controller unit tests is primarily in the handling of business logic errors. For example, if the service throws a "resource not found" exception, the controller should return a 404.

**Reply by TangoYankee** [reply]
On an experimental branch, I tried to [test mapping an exception to a status code](https://github.com/NYCPlanning/ae-zoning-api/blob/c6348af868f79e3a5ba70c47a4b5f0692c8aebc9/src/borough/borough.controller.spec.ts#L39). However, it simply fails the test because an error is thrown in the mock implementation. I am looking into whether we can use e2e tests to map exceptions to status codes.

**Reply by TangoYankee** [reply]
Update: [e2e tests can test](https://github.com/NYCPlanning/ae-zoning-api/blob/ca319ef9d42d15d3822f3f15824e841b99e909fb/test/borough.e2e-spec.ts#L36) business errors turning into status codes.

**Reply by TangoYankee** [reply]
For unit tests, the services definitely have unit-level logic that we will want to test. In the short term, this is logic like `returns a 404 when the repository returns "undefined"`. 

The controllers do not yet have unit-level logic like this. Their value is revealed at the e2e level, returning specific status codes and bodies based on the behavior of the services and repositories.

**Comment by TangoYankee** [comment]
Would [kubb-faker](https://www.kubb.dev/plugins/swagger-faker/) also be a good resource to mock results?

**Comment by TangoYankee** [comment]
We'll also want to pick an ORM before we start testing because we'll need to be mock the repos.

Here is a branch that [removes mikro orm](https://github.com/NYCPlanning/ae-zoning-api/tree/remove-mikro)

**Reply by pratishta** [reply]
Does this mean we're set on using Drizzle for our ORM?

**Reply by TangoYankee** [reply]
[we are moving forward with it for the foreseeable future](https://github.com/NYCPlanning/ae-zoning-api/discussions/116#discussioncomment-7966096)

**Comment by TangoYankee** [comment]
As a general strategy, we will want to use experiment with different implementations to learn what most make sense. In this vein, we should create research tickets for folks to experiment. They can report their findings back to this discussion. We can research, experiment, and improve with active branches.

**Comment by pratishta** [comment]
Talked with @horatiorosa and the following came up.

The "end to end" that we'd be testing would be simulating a call to an API endpoint and getting an appropriate HTTP response. The majority of the work is happening in the service (i.e. taking given bbl, fetching the data from the repository, returning the proper object). Presumably we'd want an end-to-end test for each category that we have endpoints for (e.g. borough, land uses, tax lots, etc).

When testing the service for tax lot, we want to mock the results of the repository calls so I did that in [pratishta/tax-lot-e2e](https://github.com/NYCPlanning/ae-zoning-api/blob/pratishta/tax-lot-e2e/test/tax-lot.e2e-spec.ts#L12C3-L84C5). 

This is crude but allows testing for `findTaxLotByBbl` and `findTaxLotByBblGeoJson`. 

Some thoughts / questions with this method: 
- Do want one large file of mocked a few repo response objects (e.g. tax lot, geojson of taxlot)? 
- Do want to manually come up with these responses or is there a tool we can use to fake or generate given a schema? (I guess this goes into whether we're expecting to have more API endpoints than, like, 15 or so that we currently have).

**Reply by TangoYankee** [reply]
> Do want one large file of mocked a few repo response objects (e.g. tax lot, geojson of taxlot)?

My first consideration for the file structure is for the reuse of the repo across both unit and e2e tests. My second consideration is for the scalability of the file structure as we add repos for all of the domains- boroughs, tax lots, zoning districts, etc... 

My first instinct is that each repository is a class. Correspondingly, each mock repo should only need to be a single class. We could probably have one file for each mocked repo- ie) one file for the tax lot repo mock, a second file for the borough repo mock. If we need to override the behavior of a method from its default mock behavior, that can be done within the test and against the base mock- rather than creating a new mock class for each variation in mocked behavior.

As for the placement of the files, we could probably have a folder to hold them all (exporting all the classes via an index.ts, similar to how spatial types are organized within [the drizzle-pgis module](https://github.com/NYCPlanning/ae-zoning-api/tree/main/src/drizzle-pgis))

**Reply by TangoYankee** [reply]
> Do want to manually come up with these responses or is there a tool we can use to fake or generate given a schema?

I think [Kubb/Faker](https://www.kubb.dev/plugins/swagger-faker/) is what you're looking for? (You can either play around with it or try to find some other tools)

**Comment by pratishta** [comment]
Updated [pratishta/tax-lot-e2e](https://github.com/NYCPlanning/ae-zoning-api/commit/a1992d8f4adb55652d7d6f33f0fc61165a92da6b) to include kubb/swagger-faker. Created the mock return object for repo function `findByBblSpatial` with: 
```
const taxLotGeojsonMock = {
    ...createTaxLot(),
    geometry: createMultiPolygon(),
  };
```

The issue I'm running into now is when running the test, kubb/faker not recognizing the patterns set in the openapi.yaml. 

```
● TaxLots › /GET tax-lots by geojson

    expected {
      bbl: '/^(5054146175)$/',
      borough: { id: '/\\b4\\b/', title: 'V', abbr: 'I' },
      block: 'i',
      lot: 'z',
      address: 'I',
      landUse: { id: 'R', description: 'v', color: '/^#(46Afa2c3)$/' },
      geometry: { type: 'MultiPolygon', coordinates: [ [Array] ] }
    } response body, got {
      message: 'Cannot GET /tax-lots//%5E(5054146175)$//geojson',
      error: 'Not Found',
      statusCode: 404
    }
```

I'm sure there's a way to cast or identify regex in the options but looking into that next.

**Reply by TangoYankee** [reply]
~~I'm taking a look~~ I made a branch with a working example. I'll write up my thoughts on Tuesday 
[ty/pratishta/tax-lot-e2e](https://github.com/NYCPlanning/ae-zoning-api/tree/ty/pratishta/tax-lot-e2e)

**Reply by TangoYankee** [reply]
The major two changes are to the source of the faker data and the approach to assertions.

For the faker data, I swapped swagger-faker for ["@anatine/zod-mock"](https://www.npmjs.com/package/@anatine/zod-mock). zod-mock builds fake data from zod schemas, rather than Open API schemas. From examining the mocks, it's really the repositories that we want to mock. The Open API documentation schemas are blind to the repository data; the documentation only represent the data after it's been formatted for the end-user/public API consumer. We want to mock the intermediary data, that is hidden from the end user. Aside from this point, swagger-faker also seemed to be struggling with generating fake data from regex formats. In contrast, zod-mock can create fake data from any zod schema. We will need to hand-write the zod-schema for each repository method. However, these schemas should be generally useful.  This does leave the question of making sure our repository schemas match the actual responses from the endpoints. For this, we can create a Type from each schema and mark each endpoint with the type we expect it to return.

For the assertions, we can use the zod schema generated by Kubb based on the Open API documentation; we are checking that the data follows the schema intended for the end user. In this example, we are asserting the returned data follows [the query response schema](https://github.com/NYCPlanning/ae-zoning-api/compare/main...ty/pratishta/tax-lot-e2e#diff-4d09fd082c2a280cd7fc8e42292d6d54d82c96d3aaced7f59586f4cae4c50a5cR61)

**Comment by dhochbaum-dcp** [comment]
Unit testing seems simple enough, but I receive an error with an import in a previously existing file that should not occur, which I was unable to figure out.  Code is uploaded to [123/dh-unit-test-research](https://github.com/NYCPlanning/ae-zoning-api/tree/123/dh-unit-test-research) and the error is screenshotted below.  @bmarchena reports independently having the same issue.
<img width="851" alt="image" src="https://github.com/NYCPlanning/ae-zoning-api/assets/61206501/ec424c59-de18-45c4-8767-c2e521e30848">

**Reply by TangoYankee** [reply]
@dhochbaum-dcp @bmarchena Sorry for the confusion. The ticket is to write unit tests for services. The discussion and the tickets read like they contradict each other; Tyler and I should have been clearer about the state of the discussion.

For the error shown here, my first thought is the real service is called at some point and it relies on the repository. Mocking the repository should resolve this error. The e2e test branches have an example of [mocking a repository](https://github.com/NYCPlanning/ae-zoning-api/compare/main...pratishta/tax-lot-e2e#diff-4d09fd082c2a280cd7fc8e42292d6d54d82c96d3aaced7f59586f4cae4c50a5cR92). 

The key points to mocking the repo are:
1) Do not import the real module, lest the test tries to call its dependencies
2) Create an object to mock the repository [line 92](https://github.com/NYCPlanning/ae-zoning-api/compare/main...pratishta/tax-lot-e2e#diff-4d09fd082c2a280cd7fc8e42292d6d54d82c96d3aaced7f59586f4cae4c50a5cR92)
3) Override the real repository with the mock repository [line 106](https://github.com/NYCPlanning/ae-zoning-api/compare/main...pratishta/tax-lot-e2e#diff-4d09fd082c2a280cd7fc8e42292d6d54d82c96d3aaced7f59586f4cae4c50a5cR106)

**Reply by dhochbaum-dcp** [reply]
@TangoYankee I'm unclear as to what to do.  I can import TaxLotRepository, but I am unable to override it, as done in line 106, because I get the same error as in the screenshot above.

**Reply by pratishta** [reply]
@dhochbaum-dcp, I was taking a look at your branch with Tim's changes (as of today) in mind and I think that error may have to do with the `jest-unit.json` config. There's currently no `modulePaths` included so if you have 

```
// test/jest-unit.json
{
  "moduleFileExtensions": ["js", "json", "ts"],
  "rootDir": "../",
  "modulePaths": ["<rootDir>"],
...
```
then running the unit tests will bypass the `Can't find module` error and bring about a brand new error :-D (that should be subsequently resolved with this [PR](https://github.com/NYCPlanning/ae-zoning-api/pull/140) merged)

**Reply by TangoYankee** [reply]
The lack of `modulePaths` is a correct diagnosis. However, rather than creating and maintaining a `jest-unit.json` config, we should use the jest configuration that is already in `package.json`. The linked PR also includes the change to add the modulePaths to the package.json jest config. 

The `package.json` config also enforces a naming convention of using `spec` as the test identifier, rather than `test`. This means we would want to rename tax-lot.controller.test.ts to  tax-lot.controller.spec.ts.

**Comment by TylerMatteo** [comment]
Hey everyone, thanks for all of the hard work on this during sprint A. In order to have a clear picture of what we're working on in Sprint B, let's take stock and talk next steps.

@TangoYankee - it sounds like there are still some things to be figured out with both e2e and unit tests, with e2e perhaps being a bit closer to resolution. I'm cool with carrying over the two research Issues into Sprint B, but it would be nice to have some implementation Issues created to be picked up when we feel like we have some solid testing code to build off of. Before I jump into creating Issues, I'd love your two cents. Here's how I'm thinking we can break it down:

* An Issue-per-controller file to create e2e tests. I figure we can do the zoning districts and tax lots controller for sure, and add Issues for the others as we deem necessary. I'm definitely open to doing all, since we would want that coverage eventually anyway. The requirements would specify that we want to add an e2e test for each endpoint, for each happy/unhappy path. We would exclude any `.pbf` endpoints. I'm open to also specifying endpoints or unhappy paths to exclude on a case-by-case basis if you think there are ones that would be overkill
* An Issue-per-service file for unit tests. Again, it would be great to do all but we should do tax lot and zoning district for sure. Requirements would be to add a test for each method, testing happy path and paths where exceptions are thrown.

Thoughts? I have some time to write these up this afternoon.

**Reply by TangoYankee** [reply]
The breakdown of issue per controller and service seems solid. I'm working on a branch to get my head around the specific implementation for each testing approach

**Reply by TylerMatteo** [reply]
Sounds good. I added unit and e2e Issues for tax lot, zoning district, and zoning district class services and controllers to sprint B. Their status is blocked because they're blocked by their respective research stories that will I carry over to B as well.

This maybe goes without saying but just to be explicit - the way we're breaking down this work is great for the prototyping phase we're in now but any new Issues we take on for adding new endpoints in the future shouldn't have dedicated testing Issues like this - unit and e2e testing will just be considered "definition of done" for new feature work.

**Comment by pratishta** [comment]
On the issue of refactoring the tilesets redirect by removing the injection from the controller and re-delegating it to the service and repository:

I'm wondering what the benefit of directly injecting the storage url is in the first place if we have it configured in `config.ts` as `storage`. Could we instead use `ConfigService` to get the storage url? (Going off this Nest [documentation](https://docs.nestjs.com/techniques/configuration#using-the-configservice))

```code:javascript
// src/tax-lot/tax-lot.controller.ts

@UseFilters(
  BadRequestExceptionFilter,
  InternalServerErrorExceptionFilter,
  NotFoundExceptionFilter,
)
@Controller("tax-lots")
export class TaxLotController {

  constructor(
    private readonly taxLotService: TaxLotService,
    private configService: ConfigService,
  ) {}
  ...
  ...
  ...
@Get("/:z/:x/:y.pbf")
  @Redirect()
  findTaxLotTilesets(@Param() params: { z: number; x: number; y: number }) {
    return {
      url: `${this.configService.get("storage").storageUrl}/tilesets/tax_lot/${params.z}/${params.x}/${params.y}.pbf`,
    };
  }
  ```
  and we can import the `ConfigModule` in the domain specific module.ts with:
 
```
// src/tax-lot/tax-lot.module.ts

@Module({
  imports: [ConfigModule],
  exports: [TaxLotService],
  providers: [TaxLotService, TaxLotRepository],
  controllers: [TaxLotController],
})
```
This allows the tests to pass while still having a proper endpoint for the tileset redirect. Thoughts? :0

**Reply by TangoYankee** [reply]
From the linked documentation, the benefit is strong typing.

*Screengrab of documentation*
![image](https://github.com/NYCPlanning/ae-zoning-api/assets/11068062/36763451-585d-4dff-a1de-a84c501ed45f)

I appreciate exploring alternatives for this and the information will likely inform future decisions. But, I don't want to spend too much time on this endpoint because it's [actually failing in more existential ways](https://github.com/NYCPlanning/ae-zoning-api/discussions/84) and probably going away soon.

**Reply by pratishta** [reply]
What makes me hesitant is adding more code to the service and repository files for a workaround in order to get testing to work, especially if it's going away soon anyway. In my head, I'm thinking, won't it be easier to remove `ConfigService` from the controller file vs. removing/updating the injections from both the service and repository files? And I'm thinking about the separation of concerns where the repository is meant to use drizzle to query the DB but this is a direct fetch via url. I'm not hip to the initial implementation of this endpoint nor how important it is to have the storage url strongly typed, so I will defer, but wanted to include my $0.02 about future code clean up!

**Reply by TangoYankee** [reply]
I'm comfortable interpreting the separation of the tilesets responsibilities as congruent with the general separation of concerns. ie the controller handles http, services handle business, and repositories handle data lookup. 

As far as cleanup if we need to remove the endpoint, I think the difference is trivial. It only took a few minutes to separate into different files. It should only take a few minutes to remove.
